File: tst/buffer_pct.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);
    pico_set_title("Buffer");

    Pico_Dim phy = {100,100};
    Pico_Dim log = { 10, 10};
    pico_set_view_raw(-1, &phy, NULL, &log, NULL, NULL);

    // .x.
    // xxx
    // .x.
    Pico_Color buffer[] = {
        { 0x00, 0x00, 0x00 },
        { 0xFF, 0xFF, 0x00 },
        { 0x00, 0x00, 0x00 },
        { 0xFF, 0x00, 0x00 },
        { 0x00, 0xFF, 0x00 },
        { 0x00, 0x00, 0xFF },
        { 0x00, 0x00, 0x00 },
        { 0x00, 0xFF, 0xFF },
        { 0x00, 0x00, 0x00 },
    };
    
    {
        puts("centered 3x3 on black - 1dir/1baixo");
        Pico_Rect_Pct r = { 0.5,0.5,0,0, PICO_ANCHOR_C, NULL };
        pico_output_clear();
        pico_output_draw_buffer_pct(&r, buffer, (Pico_Dim){3,3});
        //_pico_check("buf3w3h_center_black");
        pico_input_event(NULL, PICO_KEYDOWN);
    }
    {
        puts("bottomright 9x1 on white");
        Pico_Rect_Pct r = { 1,1,0,0, PICO_ANCHOR_SE, NULL };
        pico_set_color_clear((Pico_Color){0xFF, 0xFF, 0xFF});
        pico_output_clear();
        pico_output_draw_buffer_pct(&r, buffer, (Pico_Dim){9,1});
        //_pico_check("buf9w1h_rightbottom_white");
        pico_input_event(NULL, PICO_KEYDOWN);
    }

    pico_init(0);
    return 0;
}

File: lua/tst/buffer_pct.lua
require 'pico.check'

pico.init(true)
local phy = {'!', w=100, h=100}
local log = {'!', w=10, h=10}
pico.set.window { title="Buffer", dim=phy }
pico.set.view { dim=log }

do
    -- .x.
    -- xxx
    -- .x.
    local buffer = {
        {
            { r=0x00, g=0x00, b=0x00, a=0xFF },
            { r=0xFF, g=0xFF, b=0x00, a=0xFF },
            { r=0x00, g=0x00, b=0x00, a=0xFF },
        },
        {
            { r=0xFF, g=0x00, b=0x00, a=0xFF },
            { r=0x00, g=0xFF, b=0x00, a=0xFF },
            { r=0x00, g=0x00, b=0xFF, a=0xFF },
        },
        {
            pico.color.alpha('black', 0),
            { r=0x00, g=0xFF, b=0xFF, a=0xFF },
            pico.color.alpha('black', 0),
        },
    }

    print("centered 3x3 on black - 1dir/1baixo")
    local r = {'%', x=0.5, y=0.5, w=0, h=0}
    pico.output.clear()
    pico.output.draw.buffer("buf1", buffer, r)
    pico.check("buffer-01")
end

do
    local buffer = {
        {
            { r=0x00, g=0x00, b=0x00, a=0xFF },
            { r=0xFF, g=0xFF, b=0x00, a=0xFF },
            { r=0x00, g=0x00, b=0x00, a=0xFF },
            { r=0xFF, g=0x00, b=0x00, a=0xFF },
            { r=0x00, g=0xFF, b=0x00, a=0xFF },
            { r=0x00, g=0x00, b=0xFF, a=0xFF },
            pico.color.alpha('black', 0),
            { r=0x00, g=0xFF, b=0xFF, a=0xFF },
            pico.color.alpha('black', 0),
        }
    }

    print("bottomright 9x1 on white")
    pico.set.color.clear('white')
    pico.output.clear()
    local r = {'%', x=1, y=1, w=0, h=0, anchor='SE'}
    pico.output.draw.buffer("buf2", buffer, r)
    pico.check("buffer-02")
end

pico.init(false)
